crates/matrix-sdk-crypto/src/olm/account.rs

```diff
@@ -1128,32 +1122,79 @@ impl Account {
         Ok(())
     }
 
-    /// Try to decrypt an Olm message.
+    /// Try to decrypt a normal Olm message.
     ///
-    /// This try to decrypt an Olm message using all the sessions we share
+    /// This tries to decrypt an Olm message using all the sessions we share
     /// with the given sender.
-    async fn decrypt_with_existing_sessions(
+    async fn decrypt_normal_olm_message(
         store: &Store,
+        sender: &UserId,
         sender_key: Curve25519PublicKey,
         message: &OlmMessage,
-    ) -> OlmResult<Option<(Session, String)>> {
-        let s = store.get_sessions(&sender_key.to_base64()).await?;
+    ) -> OlmResult<(SessionType, String)> {
+        let OlmMessage::Normal(_) = message else {
```

Comment:
This part is a hard no from me. I prefer almost anything to this conversion of compile time guarantees to a runtime guarantee, this includes:

1. Duplicated code.
2. Big functions.
3. Complex code.
4. Dinner with in-laws.
5. Going to church.
6. Doing my taxes.
7. \<Insert your favorite unlikable thing here>